# NumPy Tutorial: A Beginner's Guide

## What is NumPy?

**NumPy** (Numerical Python) is a powerful Python library used for working with arrays and performing mathematical operations on them. It is the foundation for most data science and machine learning libraries in Python.

## Why is NumPy Important?

1. **Speed**: NumPy operations are significantly faster than regular Python lists
2. **Memory Efficiency**: NumPy arrays use less memory than Python lists
3. **Convenience**: NumPy provides easy-to-use functions for complex mathematical operations
4. **Foundation**: Libraries like Pandas, Scikit-learn, and TensorFlow are built on NumPy

## Where is NumPy Used?

- **Data Science**: Data manipulation and analysis
- **Machine Learning**: Matrix operations for neural networks
- **Scientific Computing**: Physics simulations, statistical analysis
- **Image Processing**: Images are represented as NumPy arrays
- **Financial Analysis**: Stock price analysis, risk calculations

---

# 1. Importing NumPy

Before using NumPy, we need to import it into our Python environment. The standard convention is to import NumPy with the alias `np` so we can write shorter code. Let's import NumPy now.

In [4]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/12.3 MB ? eta -:--:--
   ---------------------------------------- 0.0/12.3 MB ? eta -:--:--
   ------- -------------------------------- 2.4/12.3 MB 10.3 MB/s eta 0:00:01
   ---------------- ----------------------- 5.0/12.3 MB 11.2 MB/s eta 0:00:01
   ------------------------ --------------- 7.6/12.3 MB 11.5 MB/s eta 0:00:01
   -------------------------------- ------- 10.0/12.3 MB 11.5 MB/s eta 0:00:01
   ---------------------------------------  12.1/12.3 MB 11.6 MB/s eta 0:00:01
   ---------------------------------------- 12.3/12.3 MB 11.0 MB/s  0:00:01
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.3 -> 26.0.1
[notice] To update, run: C:\Program Files\Python312\python.exe -m pip install --upgrade pip


In [5]:
import numpy as np

**Output Explanation**: No output is displayed, but NumPy is now ready to use. We can access all NumPy functions using the `np.` prefix.

---

# 2. Python Lists vs NumPy Arrays

Before we dive into NumPy, let's understand how Python lists work. A Python list can hold different data types (integers, floats, strings) together in a single list. Let's create a Python list with mixed data types to see this flexibility.

In [6]:
A = [1, 2, 3, 4, 5.6, "7"]
A

[1, 2, 3, 4, 5.6, '7']

**Output Explanation**: The output `[1, 2, 3, 4, 5.6, '7']` shows a Python list containing integers (1, 2, 3, 4), a float (5.6), and a string ('7'). Python lists are flexible but this flexibility comes at a cost of speed and memory.

---

# 3. Creating NumPy Arrays with Dimensions

Now let's convert this Python list into a NumPy array using the `np.array()` function. We'll also use the `ndmin` parameter to specify that we want the array to have at least 3 dimensions. This will show us how NumPy handles mixed data types.

In [7]:
c = np.array(A, ndmin=3)
c

array([[['1', '2', '3', '4', '5.6', '7']]], dtype='<U32')

**Output Explanation**: The output `array([[['1', '2', '3', '4', '5.6', '7']]], dtype='<U32')` shows:
- The array has 3 dimensions (notice the three levels of brackets `[[[...]]]`)
- All values are converted to strings (dtype='<U32' means Unicode strings)
- NumPy converts all elements to a common data type (this is called upcasting)

---

Now let's verify that our variable `c` is indeed a NumPy array by checking its type using Python's built-in `type()` function.

In [8]:
type(c)

numpy.ndarray

**Output Explanation**: The output `numpy.ndarray` confirms that `c` is a NumPy N-dimensional array (ndarray).

---

# 4. Basic NumPy Array Operations

Let's create a simple 1-dimensional NumPy array with only integers. This will create a homogeneous array (all elements of the same type) and show the basic array structure.

In [9]:
b = np.array([1, 3, 4, 2, 5, 6])
b

array([1, 3, 4, 2, 5, 6])

**Output Explanation**: The output `array([1, 3, 4, 2, 5, 6])` shows a 1D array with 6 integer elements. Since all elements are integers, NumPy keeps them as integers.

---

We can find out how many elements are in our array using the `len()` function, just like with Python lists. Let's check the length of our array `b`.

In [10]:
len(b)

6

**Output Explanation**: The output `6` tells us there are 6 elements in the array.

---

# 5. Array Slicing with Negative Indices

NumPy arrays support slicing, which allows us to access portions of an array. Negative indices count from the end of the array (-1 is the last element, -2 is second to last, etc.). Let's try slicing from index -1 to -4 to understand how negative slicing works.

In [13]:
b[1:4]

array([3, 4, 2])

In [11]:
b[-1:-4]

array([], dtype=int64)

**Output Explanation**: The output `array([], dtype=int32)` is an empty array. This happens because slicing from -1 to -4 goes in the wrong direction (the default step is forward/positive), so no elements are returned. To slice backwards, we would need a negative step.

In Python, the slicing syntax is [start : stop : step]. Since you didn't provide one, Python defaults to +1. Python starts at index -1 and tries to move forward (to the right) by 1 to reach -4. But -4 is to the left of -1. Since it can't reach the destination by moving forward, it simply returns an empty array.

In [12]:
b[-1:-4:-1]

array([6, 5, 2])

# COMPARISON BETWEEN LIST AND ARRAY

## 3 Benefits of NumPy Arrays:
1. **Takes Less Memory**
2. **Faster Performance**
3. **More Convenient Operations**

Let's demonstrate each of these benefits with examples.

---

# 6. Benefit 1: Memory Efficiency

Let's compare how much memory Python lists use versus NumPy arrays. First, we need to import the `sys` module which helps us measure memory size, and the `time` module which we'll use later for speed comparisons.

In [14]:
import time
import sys

Now let's create a Python list with 2000 elements using `range()` and calculate the total memory it uses. We'll multiply the size of one integer by the number of elements to get the total memory.

In [15]:
l = range(2000)
print("Python List memory:", sys.getsizeof(2) * len(l), "bytes")

Python List memory: 56000 bytes


**Output Explanation**: Python list with 2000 elements uses 56000 bytes. Each Python integer object takes about 28 bytes of memory.

---

Now let's create a NumPy array with the same 2000 elements using `np.arange()` and calculate its memory usage. We'll use the `itemsize` property which tells us the size of each element in bytes.

In [ ]:
a = np.arange(2000)
print("NumPy Array memory:", a.itemsize * len(a), "bytes")

**Output Explanation**: NumPy array uses only 8000 bytes for the same 2000 elements - that's **7 times less memory** than the Python list (56000 vs 8000 bytes)! Each NumPy integer uses only 4 bytes compared to 28 bytes for Python integers.

---

# 7. Benefit 2: Speed (Fast Performance)

Now let's compare the speed of performing operations on Python lists versus NumPy arrays. We'll add two sequences of 100 million numbers and measure how long each approach takes. First, let's test with Python lists using a list comprehension and the `zip()` function.

In [ ]:
a1 = range(100000000)
b1 = range(100000000)

start = time.time()
c1 = [(x + y) for x, y in zip(a1, b1)]
print("Python List time:", (time.time() - start), "seconds")

**Output Explanation**: Python lists took approximately **10.5 seconds** to add 100 million pairs of numbers. That's quite slow for modern computers!

---

Now let's perform the same operation using NumPy arrays. We'll create two arrays with `np.arange()` and simply add them using the `+` operator. NumPy handles the element-wise addition internally.

In [ ]:
a2 = np.arange(100000000)
b2 = np.arange(100000000)

start = time.time()
c2 = a2 + b2
print("NumPy Array time:", (time.time() - start), "seconds")

**Output Explanation**: NumPy took only approximately **0.15 seconds** - that's about **70 times faster** than Python lists! This dramatic speed improvement is why NumPy is essential for data science and scientific computing.

---

# 8. Benefit 3: Convenience

NumPy makes mathematical operations much simpler and more intuitive. Let's create two arrays of 100 elements each and add them together. With NumPy, we can simply use the `+` operator without writing any loops.

In [ ]:
n1 = np.arange(100)
n2 = np.arange(100)
n1 + n2

**Output Explanation**: The output shows element-wise addition: `[0+0, 1+1, 2+2, ...]` = `[0, 2, 4, 6, ...]`. With NumPy, we simply use `+` instead of writing loops!

---

Now let's see how the same operation would be done with Python lists. We need to create empty list, loop through both lists using `zip()`, add each pair of elements, and append the result. This requires much more code.

In [ ]:
a = [1, 2, 3, 4, 5, 6]
b = [1, 2, 3, 4, 5, 6]

d = []
for i, j in zip(a, b):
    c = i + j
    d.append(c)
d

**Output Explanation**: We get the same result `[2, 4, 6, 8, 10, 12]`, but it required 5 lines of code compared to just `n1 + n2` with NumPy! This shows how convenient NumPy is for mathematical operations.

---

# 9. Array Dimensions (ndim and ndmin)

NumPy arrays can have multiple dimensions. The `ndmin` parameter lets us specify the minimum number of dimensions when creating an array. Let's create a 3-dimensional array from a simple list and check its dimensions using the `.ndim` property.

In [ ]:
b = np.array([1, 2, 3], ndmin=3)
print("Array:", b)
print("Dimensions:", b.ndim)

**Output Explanation**: The array `[[[1, 2, 3]]]` has 3 dimensions (notice 3 levels of brackets). The `.ndim` property returns `3`, confirming the number of dimensions.

---

# 10. Data Type Conversion (dtype - Upcasting)

We can specify the data type of array elements using the `dtype` parameter. NumPy will convert all elements to the specified type. Let's create an array of integers and convert them to strings using `dtype="U3"` (Unicode string with max length 3).

In [ ]:
w = np.array([2, 3, 5], dtype="U3")
w

**Output Explanation**: The output `array(['2', '3', '5'], dtype='<U3')` shows that all integers have been converted to strings. The dtype '<U3' means Unicode strings with maximum length of 3 characters.

---

Now let's convert integers to floating-point numbers using `dtype="<f4"` which means 4-byte (32-bit) float. This is useful when we need decimal precision in calculations.

In [ ]:
c = np.array([2, 9, 100, 234, 453, 9], dtype="<f4")
c

**Output Explanation**: The output shows all integers converted to floats (notice the decimal points). The dtype '<f4' represents 32-bit floating-point numbers.

---

# 11. Array Properties: shape, size, ndim, itemsize

NumPy arrays have several useful properties that tell us about their structure. Let's create a 2D array and explore these properties: `shape` (dimensions), `size` (total elements), `ndim` (number of dimensions), and `itemsize` (bytes per element).

In [ ]:
a = np.array([[1, 2.2, 3], [4, 5, 6]], dtype="U2")
print("Array:\n", a)
print("\nShape (rows, cols):", a.shape)
print("Total elements:", a.size)
print("Dimensions:", a.ndim)
print("Bytes per element:", a.itemsize)

**Output Explanation**:
- `shape (2, 3)`: The array has 2 rows and 3 columns
- `size 6`: Total of 6 elements (2 × 3 = 6)
- `ndim 2`: This is a 2-dimensional array
- `itemsize`: Number of bytes each element occupies in memory

---

# 12. Shallow Copy (Reference Assignment)

Understanding how copying works is crucial to avoid bugs. When you assign a list or array to a new variable using `=`, both variables point to the same data in memory. Let's demonstrate this by creating a list, assigning it to another variable, and modifying one of them.

In [ ]:
l = [1, 2, 3, 4, 5]
l1 = l  # Both point to same memory
l1[1] = 100  # Modify l1

print("l1:", l1)
print("l:", l)
print("Same memory?", id(l) == id(l1))

**Output Explanation**: Both `l` and `l1` show `[1, 100, 3, 4, 5]`. Changing `l1` also changed `l` because they both reference the same memory location! The `id()` function confirms they have the same memory address.

---

# 13. Using copy() for Independent Copies

To create an independent copy that doesn't affect the original, we use the `.copy()` method. Let's create a list, make a copy of it, and then modify the copy to see that the original remains unchanged.

In [ ]:
l = [1, 2, 3, 4, 5]
l2 = l.copy()  # Creates independent copy
l2[1] = 100  # Modify l2

print("l2:", l2)
print("l:", l)
print("Same memory?", id(l) == id(l2))

**Output Explanation**: Now `l2` shows `[1, 100, 3, 4, 5]` but `l` remains `[1, 2, 3, 4, 5]`. They are independent! The different memory IDs confirm they are separate objects.

---

# 14. Deep Copy for Nested Structures

The `.copy()` method only creates a shallow copy, which means nested objects (like lists inside lists) still share memory. For completely independent copies of nested structures, we need `deepcopy()` from the `copy` module. Let's demonstrate this with a nested list.

In [ ]:
import copy

l = [[1, 2, 3], [4, 5, [55, 66, 67], 4, 6]]
l2 = copy.deepcopy(l)  # Creates completely independent copy
l2[1][2][1] = 100  # Modify deeply nested element

print("l2:", l2)
print("l:", l)

**Output Explanation**: `l2` is modified (66 changed to 100) but `l` remains unchanged. The `deepcopy()` function creates completely independent copies of all nested objects, no matter how deep!

---

# 15. np.arange() - Create Arrays with Step Value

The `np.arange()` function is similar to Python's `range()` but can handle floating-point step values. Let's create an array starting from 0, going up to (but not including) 20, with a step of 2.5.

In [22]:
l = range(0,20,2)  # range(start (optional), stop (req), step (optional)). range() function generates an immutable sequence of integers.
# start defaults to 0. 
# step defaults to 1. cannot be zero (throws error). Negative value can be used to count backward. 
# l = range(0,20,0)  -- throws error

# l = range(0,20,2.5) -- throws error
# print (l) -- print range(0,20,2) as it is since range returns an iterable not exeuted on memory.  It is memory-efficient because 
# it generates the numbers on demand (lazy evaluation), rather than storing the entire sequence in memory
print([i for i in l])

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18]


In [19]:
np.arange(0, 20, 2.5)
# np.arange(0, 20) -- creates 20 elements from 0 to 19 if no step is provided. 

array([ 0. ,  2.5,  5. ,  7.5, 10. , 12.5, 15. , 17.5])

**Output Explanation**: The output `array([ 0. ,  2.5,  5. ,  7.5, 10. , 12.5, 15. , 17.5])` shows values from 0 to 17.5 with step 2.5. Unlike Python's `range()`, NumPy's `arange()` supports float step values!

---

We can also count backwards by using a negative step value. Let's create an array that counts down from 50 to 0 with a step of -5.

In [ ]:
np.arange(50, -1, -5)

**Output Explanation**: The output `array([50, 45, 40, 35, 30, 25, 20, 15, 10,  5,  0])` shows counting backwards from 50 to 0 with step -5.

---

# 16. np.linspace() - Create Arrays with Specific Number of Points

While `arange()` uses a step value, `linspace()` lets us specify exactly how many points we want between two values. Let's create an array of 10 evenly spaced values between 1 and 10.

In [ ]:
np.linspace(1, 10, 10)

**Output Explanation**: The output shows exactly 10 evenly spaced values from 1 to 10, with both endpoints included.

---

We can also see the step size that `linspace()` calculated by using `retstep=True`. This returns a tuple containing the array and the step value. Let's create 5 values between 2.0 and 3.0 and see the step size.

In [ ]:
np.linspace(2.0, 3.0, num=5, retstep=True)

**Output Explanation**: Returns a tuple: the array `[2., 2.25, 2.5, 2.75, 3.]` and the step size `0.25`. This is useful when you need to know the exact spacing between values.

---

# 17. Reshaping Arrays

The `reshape()` method changes the dimensions of an array without changing its data. The total number of elements must remain the same. Let's create a 1D array of 32 elements and reshape it into a 5-dimensional array (2×2×2×2×2 = 32).

In [ ]:
a = np.arange(32).reshape(2, 2, 2, 2, 2)
print("Shape:", a.shape)
print("Dimensions:", a.ndim)
a

**Output Explanation**: The 32 elements are arranged as a 5-dimensional array. The shape `(2, 2, 2, 2, 2)` means each dimension has 2 elements, and `ndim` confirms 5 dimensions.

---

Let's create another example: reshaping numbers 1-20 into a 3D array with 2 blocks, each containing 5 rows and 2 columns.

In [ ]:
x = np.arange(1, 21).reshape(2, 5, 2)
x

**Output Explanation**: Numbers 1-20 are reshaped into a 3D array: 2 blocks, each with 5 rows and 2 columns (2 × 5 × 2 = 20 elements).

---

# 18. Creating Special Matrices

NumPy provides convenient functions to create commonly used matrices. Let's create a 4×4 matrix filled with zeros using `np.zeros()`. This is useful for initializing matrices before filling them with data.

In [ ]:
print("Zeros matrix (4x4):")
print(np.zeros((4, 4)))

Now let's create a 2×3 matrix filled with ones using `np.ones()`. This is often used as a starting point for calculations.

In [ ]:
print("Ones matrix (2x3):")
print(np.ones((2, 3)))

We can create a matrix filled with any value by multiplying the ones matrix by that value. Let's create a 3×4 matrix filled with 5s.

In [ ]:
print("Matrix filled with 5s (3x4):")
print(5 * np.ones((3, 4)))

The identity matrix is a special square matrix with 1s on the diagonal and 0s elsewhere. It's fundamental in linear algebra. Let's create a 5×5 identity matrix using `np.eye()`.

In [ ]:
print("Identity matrix (5x5):")
print(np.eye(5))

**Output Explanation**: 
- `zeros()`: Creates a matrix filled with zeros
- `ones()`: Creates a matrix filled with ones
- `eye()`: Creates an identity matrix (ones on diagonal, zeros elsewhere)

---

# 19. Diagonal Operations

NumPy provides functions to work with diagonal elements of matrices. Let's create a 5×5 matrix and extract its diagonal elements using `np.diag()`.

In [ ]:
x = np.arange(25).reshape(5, 5)
print("Original matrix:")
print(x)
print("\nDiagonal elements:", np.diag(x))

**Output Explanation**: The `np.diag()` function extracts the diagonal elements (0, 6, 12, 18, 24) from the 5×5 matrix.

---

We can also create a diagonal matrix from a 1D array. By applying `np.diag()` twice, we first extract the diagonal and then create a new matrix with only those diagonal elements.

In [ ]:
np.diag(np.diag(x))

**Output Explanation**: This creates a new matrix with only the diagonal values (0, 6, 12, 18, 24) and zeros everywhere else.

---

The `np.diagflat()` function creates a 2D array with the flattened input as the diagonal. Let's use it to create a diagonal matrix from a nested list.

In [ ]:
np.diagflat([[1, 2, 10], [3, 4, 10]])

**Output Explanation**: The `diagflat()` function flattens the input `[[1,2,10],[3,4,10]]` to `[1,2,10,3,4,10]` and places these values on the diagonal of a new matrix.

---

# 20. Random Number Generation

NumPy has powerful random number generation capabilities. Let's first import the `randint` function with an alias `ri` for convenience. Then we'll generate a random 4-digit number between 1000 and 9999.

In [ ]:
from numpy.random import randint as ri

print("Random 4-digit number:", np.random.randint(1000, 9999))

Now let's generate random numbers from a standard normal distribution (mean=0, standard deviation=1) using `np.random.randn()`. We'll create a 2×4 matrix of such random values.

In [ ]:
print("Random 2x4 matrix (normal distribution):")
print(np.random.randn(2, 4))

We can also generate a matrix of random integers within a specific range. Let's create a 5×4 matrix with random integers between 100 and 999.

In [ ]:
print("Random 5x4 integer matrix (100-999):")
print(np.random.randint(100, 1000, (5, 4)))

**Output Explanation**: 
- `randint(low, high)`: Generates a random integer in the range [low, high)
- `randn(rows, cols)`: Generates random values from standard normal distribution
- `randint(low, high, shape)`: Generates a matrix of random integers

---

# 21. Fancy Indexing

Fancy indexing allows us to access multiple elements at once using a list of indices. Let's create a 5×2 array and then select specific rows (rows 0, 2, and 3) using fancy indexing.

In [ ]:
arr = np.arange(10, 20).reshape(5, 2)
print("Original array:")
print(arr)
print("\nRows 0, 2, and 3:")
print(arr[[0, 2, 3]])

**Output Explanation**: Using `arr[[0, 2, 3]]` returns rows at indices 0, 2, and 3. This is much more flexible than basic slicing.

---

We can also use fancy indexing on 1D arrays. Let's create an array of even numbers and select elements at specific indices (2, 4, and 9).

In [ ]:
arr = np.arange(0, 21, 2)
print("Array:", arr)
print("Elements at indices 2, 4, 9:", arr[[2, 4, 9]])

**Output Explanation**: The array contains even numbers 0-20. Using `arr[[2, 4, 9]]` returns elements at those indices: 4, 8, and 18.

---

# 22. Boolean Subsetting (Filtering)

Boolean subsetting allows us to select elements based on conditions. Let's create a random 3×5 matrix and then filter it to show only elements greater than 20.

In [ ]:
mat = np.array(ri(10, 100, 15)).reshape(3, 5)
print("Matrix:")
print(mat)

Now let's use a boolean condition to filter the matrix. We'll select all elements that are greater than 20.

In [ ]:
print("Elements > 20:", mat[mat > 20])

We can combine multiple conditions using `&` (AND) and `|` (OR) operators. Let's find elements that are both greater than 50 AND even numbers.

In [ ]:
print("Elements > 50 AND even:", mat[(mat > 50) & (mat % 2 == 0)])

**Output Explanation**: 
- `mat[mat > 20]`: Returns all elements greater than 20 as a 1D array
- `mat[(mat > 50) & (mat % 2 == 0)]`: Returns elements that satisfy both conditions
- Use `&` for AND and `|` for OR when combining conditions

---

# 23. Universal Functions (Array Arithmetic)

Universal functions (ufuncs) perform element-wise operations on arrays. Let's create two 3×3 matrices with random integers and demonstrate various arithmetic operations.

In [ ]:
mat1 = np.array(ri(1, 10, 9)).reshape(3, 3)
mat2 = np.array(ri(1, 10, 9)).reshape(3, 3)

print("Matrix 1:")
print(mat1)
print("\nMatrix 2:")
print(mat2)

Let's add the two matrices element by element. Each element in mat1 will be added to the corresponding element in mat2.

In [ ]:
print("Addition (mat1 + mat2):")
print(mat1 + mat2)

Now let's multiply the matrices element by element. Note: This is NOT matrix multiplication, but element-wise multiplication.

In [ ]:
print("Element-wise multiplication (mat1 * mat2):")
print(mat1 * mat2)

Let's divide mat1 by mat2 element by element. The result will be floating-point numbers.

In [ ]:
print("Division (mat1 / mat2):")
print(mat1 / mat2)

We can also create linear combinations of matrices. Let's calculate 3 times mat1 minus 2 times mat2.

In [ ]:
print("Linear combination (3*mat1 - 2*mat2):")
print(3 * mat1 - 2 * mat2)

Scalar operations apply to every element. Let's add 100 to every element in mat1.

In [ ]:
print("Scalar addition (100 + mat1):")
print(100 + mat1)

We can also apply mathematical functions element-wise. Let's calculate the square root of each element in mat1 by raising it to the power of 0.5.

In [ ]:
print("Square root (mat1 ** 0.5):")
print(mat1 ** 0.5)

**Output Explanation**: All arithmetic operations work element-by-element. Scalar values (like 100 or 3) are automatically broadcast to all elements of the array.

---

# 24. Broadcasting

Broadcasting is a powerful feature that allows NumPy to perform operations between arrays of different shapes. Let's create a 4×4 matrix of zeros and a 1D array of 4 elements, then add them together.

In [ ]:
start = np.zeros((4, 4))
add_rows = np.array([1, 0, 2, 3])

print("Zeros matrix (4x4):")
print(start)
print("\n1D array:", add_rows)
print("\nAfter broadcasting (start + add_rows):")
print(start + add_rows)

**Output Explanation**: The 1D array `[1, 0, 2, 3]` is "broadcast" (replicated) to match the shape of the 4×4 matrix. Each row of the result becomes `[1, 0, 2, 3]`. This is broadcasting in action!

---

# 25. Stacking Arrays (vstack and hstack)

NumPy provides functions to combine arrays. Let's create a 3×3 matrix and a row vector, then demonstrate both vertical and horizontal stacking.

In [ ]:
a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
b = np.array([[10, 20, 30]])

print("Matrix a:")
print(a)
print("\nRow vector b:", b)

Let's use `vstack()` to stack the arrays vertically (add rows). This will add the row vector as a new row below the matrix.

In [ ]:
print("Vertical stack (vstack):")
print(np.vstack((a, b)))

**Output Explanation**: The `vstack()` function stacks arrays vertically, adding the row `[10, 20, 30]` as a new bottom row to matrix `a`.

---

For horizontal stacking, we need a column vector. Let's transpose `b` to convert it from a row to a column vector using `.T`, then use `hstack()` to add it as a new column.

In [ ]:
c = b.T  # Transpose to column vector
print("Column vector c (transposed b):")
print(c)
print("\nHorizontal stack (hstack):")
print(np.hstack((a, c)))

**Output Explanation**: 
- `.T` transposes the array, converting rows to columns (and vice versa)
- `hstack()` stacks arrays horizontally, adding the column `[10, 20, 30]` to the right side of matrix `a`

---

# Summary

In this tutorial, we covered:

1. **What is NumPy** - A powerful library for numerical computing
2. **Why NumPy** - Memory efficiency, speed, and convenience
3. **Creating Arrays** - `np.array()`, `np.arange()`, `np.linspace()`
4. **Array Properties** - `shape`, `size`, `ndim`, `dtype`, `itemsize`
5. **Upcasting** - Automatic type conversion
6. **Shallow vs Deep Copy** - Understanding references and copies
7. **Reshaping** - Changing array dimensions
8. **Special Matrices** - `zeros()`, `ones()`, `eye()`, `diag()`
9. **Random Numbers** - `randint()`, `randn()`
10. **Indexing** - Basic, fancy, and boolean indexing
11. **Universal Functions** - Element-wise operations
12. **Broadcasting** - Operations on arrays of different shapes
13. **Stacking** - `vstack()`, `hstack()`

NumPy is the foundation of data science in Python. Mastering these concepts will help you work with Pandas, Matplotlib, Scikit-learn, and other powerful libraries!